src/d3a/models/market/two_sided.py

```diff
@@ -257,35 +255,57 @@ def accept_bid_offer_pair(self, bid, offer, clearing_rate, trade_bid_info, selec
     def match_recommendations(
             self, recommendations: List[BidOfferMatch.serializable_dict]) -> None:
         """Match a list of bid/offer pairs, create trades and residual offers/bids."""
-        if recommendations is None:
-            return
-        for index, recommended_pair in enumerate(recommendations):
+
+        while recommendations:
+            recommended_pair = recommendations.pop(0)
             selected_energy = recommended_pair["selected_energy"]
-            market_offer = self.offers.get(recommended_pair["offer"]["id"])
-            market_bid = self.bids.get(recommended_pair["bid"]["id"])
-
-            if not market_offer:
-                raise OfferNotFoundException
-            if not market_bid:
-                raise BidNotFoundException
-
-            original_bid_rate = market_bid.original_bid_price / market_bid.energy
-            trade_bid_info = TradeBidOfferInfo(
-                original_bid_rate=original_bid_rate,
-                propagated_bid_rate=market_bid.price/market_bid.energy,
-                original_offer_rate=market_offer.original_offer_price/market_offer.energy,
-                propagated_offer_rate=market_offer.price/market_offer.energy,
-                trade_rate=original_bid_rate)
-
-            bid_trade, trade = self.accept_bid_offer_pair(
-                market_bid, market_offer, recommended_pair["trade_rate"],
-                trade_bid_info, selected_energy
-            )
-
-            if trade.residual is not None or bid_trade.residual is not None:
-                recommendations = self._replace_offers_bids_with_residual_in_matching_list(
-                    recommendations, index+1, trade, bid_trade
+            clearing_rate = recommended_pair["trade_rate"]
+            market_offers = [
+                self.offers.get(offer["id"]) for offer in recommended_pair["offers"]]
+            market_bids = [self.bids.get(bid["id"]) for bid in recommended_pair["bids"]]
+
+            if not all(market_offers):
+                # If not all received offers exist in the market, skip the current recommendation
+                continue
+            if not all(market_bids):
+                # If not all received bids exist in the market, skip the current recommendation
+                continue
+
+            self.validate_bid_offer_match(
+                market_bids, market_offers,
+                clearing_rate, selected_energy)
+
+            market_offers = iter(market_offers)
+            market_bids = iter(market_bids)
+            market_offer = next(market_offers)
+            market_bid = next(market_bids)
+            while True:
+                original_bid_rate = market_bid.original_bid_price / market_bid.energy
+                trade_bid_info = TradeBidOfferInfo(
+                    original_bid_rate=original_bid_rate,
+                    propagated_bid_rate=market_bid.price / market_bid.energy,
```

Comment:
Trivial, but you can use `energy_rate`. Same goes for 2 lines below for the propagated_offer_rate